File: sampleCode/CSharp/ConsoleApp/Endpoints/ProjectEndpoints.cs
namespace ConsoleApp.Endpoints;

/*
# MIT License

# Copyright (c) 2023 IMPLAN

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
*/

public sealed record class Project
{
    /// <summary>
    /// The unique identifier for this Project
    /// </summary>
    public Guid Id { get; set; }
    /// <summary>
    /// The unique description for this Project
    /// </summary>
    public required string Title { get; set; }
    /// <summary>
    /// The Aggregation Scheme this Project is using
    /// </summary>
    public required int AggregationSchemeId { get; set; }
    /// <summary>
    /// The Household Set this Project is using
    /// </summary>
    public required int HouseholdSetId { get; set; }
    /// <summary>
    /// Whether or not this Project is using Multi-Region Input/Output (MRIO) Analysis
    /// </summary>
    public bool IsMrio { get; set; } = false;
    /// <summary>
    /// If present, the identifier of the Folder that the Project is located under in IMPLAN Cloud
    /// </summary>
    public int? FolderId { get; set; } = null;
    /// <summary>
    /// If an Impact Analysis has already been performed for this Project, the Id of the last one (used for querying Analysis status)
    /// </summary>
    public long? LastImpactRunId { get; set; } = null;
}

public class ProjectEndpoints
{
    public static Project Create(Project project)
    {
        RestRequest request = new RestRequest("api/v1/impact/project");
        request.Method = Method.Post;
        
        // The body must be the Jsonified Project model
        request.AddJsonBody(project);

        return Rest.GetResponseData<Project>(request).ThrowIfNull();
    }

    public static Project GetProject(Guid projectGuid)
    {
        RestRequest request = new RestRequest("api/v1/impact/project/{projectGuid}");
        request.Method = Method.Get;
        request.AddUrlSegment("projectGuid", projectGuid);

        return Rest.GetResponseData<Project>(request).ThrowIfNull();
    }

    /// <summary>
    /// Returns a list of <see cref="Project"/>s that belong to the currently Authorized Implan User
    /// </summary>
    /// <returns></returns>
    public static Project[] GetProjects()
    {
        RestRequest request = new RestRequest("api/v1/impact/project");
        request.Method = Method.Get;

        return Rest.GetResponseData<Project[]>(request).ThrowIfNull();
    }

    /// <summary>
    /// Returns a list of <see cref="Project"/>s that have been shared with the currently Authorized Implan User
    /// </summary>
    /// <returns></returns>
    public static Project[] GetSharedProjects()
    {
        RestRequest request = new RestRequest("api/v1/impact/project/shared");
        request.Method = Method.Get;

        return Rest.GetResponseData<Project[]>(request).ThrowIfNull();
    }
}

File: sampleCode/R/R - Object Oriented/Endpoints/ProjectEndpoints.R
# MIT License

# Copyright (c) 2023 IMPLAN

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

source(auth_env$Rest)
source(auth_env$Extensions)

# Define the Project class
setClass("Project",
         slots = list(
           Id = "character",
           Title = "character",
           AggregationSchemeId = "integer",
           HouseholdSetId = "integer",
           IsMrio = "logical",
           FolderId = "integer",
           LastImpactRunId = "integer"
         ))


# Define the ProjectEndpoints class which contains Project
setClass("ProjectEndpoints",
         contains = "Project")

# Create a generic function for Create
setGeneric("Create", function(object) standardGeneric("Create"))

# Define a method for the Create function with the ProjectEndpoints class
setMethod("Create", signature = "ProjectEndpoints", function(object){
  #url_object <- new("Rest", baseurl = auth_env$baseurl, lastSet = Sys.time())
  #url <- auth_env$createproject_url
  
  url <- "api/v1/impact/project"
  
  print(url)
  
  project_list_body <- convert_slots_into_list(object)
  
  # Convert NA to NULL in event1_list
  project_list_body <- lapply(project_list_body, convert_na_to_null)
  
  # Print the updated event1_list
  print(project_list_body)
  
  title <- class(object)
  
  # Make the POST request
  project_creation <- ThrowIfNull(extension,GetResponseData(rest, url = url, method_type = "POST", title_instances = title, response_body = project_list_body))
  
  return(project_creation)
})


# Create a generic function for Create
setGeneric("GetProject", function(object,ProjectId) standardGeneric("GetProject"))

# Define a method for the Create function with the ProjectEndpoints class
setMethod("GetProject", signature = "ProjectEndpoints", function(object,ProjectId){
  
  print(paste0("I came here with",ProjectId))

  GetProject_url <- paste0("api/v1/impact/project/",ProjectId)
  
  title <- class(object)

  GetProject <- ThrowIfNull(extension,GetResponseData(rest, url = GetProject_url, method_type = "GET",title_instances = title))

  return(GetProject)

})


# Create a generic function for Create
setGeneric("GetProjects", function(object) standardGeneric("GetProjects"))

# Define a method for the Create function with the ProjectEndpoints class
setMethod("GetProjects", signature = "ProjectEndpoints", function(object){

  GetProjects_url <- "api/v1/impact/project"
  
  title <- class(object)

  GetProjects <- ThrowIfNull(extension,GetResponseData(rest, url = GetProjects_url, method_type = "GET", title_instances = title))

  return(GetProjects)

})


# Create a generic function for Create
setGeneric("GetSharedProjects", function(object) standardGeneric("GetSharedProjects"))

# Define a method for the Create function with the ProjectEndpoints class
setMethod("GetSharedProjects", signature = "ProjectEndpoints", function(object){

  GetSharedProjects_url <- "api/v1/impact/project/shared"
  
  title <- class(object)

  GetSharedProjects <- ThrowIfNull(extension,GetResponseData(rest, url = GetSharedProjects_url, method_type = "GET", title_instances = title))

  return(GetSharedProjects)

})
